quinn-proto/src/cid_generator.rs

```diff
@@ -86,3 +90,93 @@ impl ConnectionIdGenerator for RandomConnectionIdGenerator {
         self.lifetime
     }
 }
+
+/// Generates 8-byte connection IDs that can be efficiently
+/// [`validate`](ConnectionIdGenerator::validate)d
+pub struct KeyedConnectionIdGenerator {
+    key: Arc<dyn crypto::HmacKey>,
+    lifetime: Option<Duration>,
+}
+
+impl KeyedConnectionIdGenerator {
+    /// Create a generator with a random key
+    #[cfg(feature = "ring")]
+    pub fn new() -> Self {
+        let mut reset_key = [0; 64];
+        rand::thread_rng().fill_bytes(&mut reset_key);
+
+        let key = Arc::new(ring::hmac::Key::new(ring::hmac::HMAC_SHA256, &reset_key));
+        Self::from_key(key)
+    }
+
+    /// Create a generator with a specific key
+    pub fn from_key(key: Arc<dyn crypto::HmacKey>) -> Self {
+        assert!(
+            key.signature_len() < MAX_SIGNATURE_LEN,
+            "key must generate at most a 128 byte signature"
+        );
+        Self {
+            key,
+            lifetime: None,
+        }
+    }
+
+    /// Set the lifetime of CIDs created by this generator
+    pub fn set_lifetime(&mut self, d: Duration) -> &mut Self {
+        self.lifetime = Some(d);
+        self
+    }
+}
+
+#[cfg(feature = "ring")]
+impl Default for KeyedConnectionIdGenerator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ConnectionIdGenerator for KeyedConnectionIdGenerator {
+    fn generate_cid(&mut self) -> ConnectionId {
+        let mut bytes_arr = [0; NONCE_LEN + MAX_SIGNATURE_LEN];
+        rand::thread_rng().fill_bytes(&mut bytes_arr[..NONCE_LEN]);
+        let (nonce, signature) = bytes_arr.split_at_mut(NONCE_LEN);
+        self.key
+            .sign(nonce, &mut signature[..self.key.signature_len()]);
+        ConnectionId::new(&bytes_arr[..self.cid_len()])
+    }
+
+    fn cid_len(&self) -> usize {
+        NONCE_LEN + SIGNATURE_LEN
+    }
+
+    fn cid_lifetime(&self) -> Option<Duration> {
+        self.lifetime
+    }
+
+    fn validate(&self, cid: &ConnectionId) -> Result<(), InvalidCid> {
+        let (nonce, signature) = cid.split_at(NONCE_LEN);
+        let mut expected_signature = [0; MAX_SIGNATURE_LEN];
+        self.key
+            .sign(nonce, &mut expected_signature[..self.key.signature_len()]);
+        (expected_signature[..SIGNATURE_LEN] == signature[..])
+            .then_some(())
+            .ok_or(InvalidCid)
```

Comment:
Style: don't love the `then_some(()).ok_or(InvalidCid)` chain. Explicit `match` at the cost of one more line?

```rust
match expected_signature[..SIGNATURE_LEN] == signature[..] {
     true => Ok(()),
     false => Err(InvalidCid),
}
```